src/infra/core/src/push_status_service_impl.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use dill::{component, interface};
+use internal_error::{InternalError, ResultIntoInternal};
+use kamu_core::{
+    BlockRef,
+    DatasetFactory,
+    DatasetPushStatuses,
+    DatasetRepository,
+    PushStatusService,
+    RemoteAliasKind,
+    RemoteAliasesRegistry,
+    SyncStatus,
+};
+use opendatafabric::{DatasetHandle, DatasetRefRemote};
+
+#[component(pub)]
+#[interface(dyn PushStatusService)]
+pub struct PushStatusServiceImpl {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    dataset_factory: Arc<dyn DatasetFactory>,
+    remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+}
+
+impl PushStatusServiceImpl {
+    pub fn new(
+        dataset_repo: Arc<dyn DatasetRepository>,
+        dataset_factory: Arc<dyn DatasetFactory>,
+        remote_alias_reg: Arc<dyn RemoteAliasesRegistry>,
+    ) -> Self {
+        Self {
+            dataset_repo,
+            dataset_factory,
+            remote_alias_reg,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl PushStatusService for PushStatusServiceImpl {
+    async fn get_statuses(
+        &self,
+        dataset_handle: &DatasetHandle,
+    ) -> Result<DatasetPushStatuses, InternalError> {
+        let local_head = self
+            .dataset_repo
+            .get_dataset_by_handle(dataset_handle)
+            .as_metadata_chain()
+            .resolve_ref(&BlockRef::Head)
+            .await
+            .int_err()?;
+
+        let aliases = self
+            .remote_alias_reg
+            .get_remote_aliases(&dataset_handle.as_local_ref())
+            .await
+            .int_err()?;
+        let push_aliases: Vec<&DatasetRefRemote> =
+            aliases.get_by_kind(RemoteAliasKind::Push).collect();
+
+        let mut statuses = HashMap::new();
+
+        for alias in push_aliases {
+            let maybe_ds = match alias.url() {
+                None => None,
+                Some(url) => self.dataset_factory.get_dataset(url, false).await.ok(),
+            };
+            let status = match maybe_ds {
+                None => None,
+                Some(ds) => ds
+                    .as_metadata_chain()
+                    .resolve_ref(&BlockRef::Head)
+                    .await
+                    .ok()
+                    .map(|remote_head| {
+                        if remote_head == local_head {
+                            SyncStatus::InSync
+                        } else {
+                            SyncStatus::OutOfSync
+                        }
```

Comment:
I wonder if it makes any sense to differentiate between the cases:
 - local dataset is ahead
 - remote dataset is ahead
 - local & remote have diverged
 
 We have something like this in `SyncService` where we compare chains.